### PR TITLE
Panache check

### LINKED ISSUE #5433: Be more explicit in case of provided id and usage of PanacheMongoEntity
**Description**
`PanacheMongoEntity` should be used when a user don't want to handle itself the MongoDB `_id` field. If the user wants to handle the field itself, he should use the `PanacheMongoEntityBase` instead.

When a user misuse the `PanacgeMongoEntity` and specify it's own `_id` with a `@BsonId` annotation, this generates a cryptic error : `readObjectId can only be called when CurrentBSONType is OBJECT_ID, not when CurrentBSONType is INT64.`

See the following zulip thread: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/mongo.20panache.20-.20throw.20missing.20implementation

**Implementation ideas**
We should detect such mis-usage of `PanacheMongoEntity` and provides a meaninful error to the user.

### PR DESCRIPTION
**Pull request by loicmathieu**
Fixes #5433

This PR adds a check to crash the build if an ID field is defined when using PanacheEntity or PanacheMongoEntity.

I check the performance impact on the build and it's less than 5ms for the integration test.

### CODE REVIEW

**Comment by loicmathieu** [comment]
/cc @FroMage @emmanuelbernard 

This is the check that we discussed recently to avoid having an entity redefining its ID.

File: extensions/panache/hibernate-orm-panache/deployment/pom.xml
Comment (gastaldi) [inline extensions/panache/hibernate-orm-panache/deployment/pom.xml:80]: Where is this dependency used? I wonder if we should add to `quarkus-junit5-internal` instead?

Comment (gastaldi) [reply extensions/panache/hibernate-orm-panache/deployment/pom.xml:80]: Also fixing the version is a bad idea, the dependency is already declared in the `junit-bom` that is imported in the runtime BOM.

File: extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java:26]: Hmm, I think we do not need to specify the configuration file since it is `application.properties` so it'll be picked by default.

File: extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java:52]: I wonder if in this test case we could have used this pattern instead https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/sql_load_script/InvalidMultilineSqlLoadScriptTestCase.java#L17 ?

File: extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdMongoEntityTest.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdMongoEntityTest.java:35]: see this comment https://github.com/quarkusio/quarkus/pull/8549/files#r407481562 and let me know what you think

File: extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java:52]: Yes, this was even documented in the extionsion guide, I need to re-read it carefully ;)

> If you want to test that an extension properly fails at build time, use the setExpectedException method:

Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java]: I updated the test with `.setExpectedException(BuildException.class)` instead of using the Launcher

File: extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/test/java/io/quarkus/mongodb/panache/DuplicateIdReactiveMongoEntityTest.java]: True: this was copied from an existing test but not needed.
Removed.

File: extensions/panache/hibernate-orm-panache/deployment/pom.xml
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/deployment/pom.xml]: I used `.setExpectedException(BuildException.class)` instead of the JUnit platform Launcher so this is no longuer needed.

**Comment by loicmathieu** [comment]
In PR #8576 I saw that a `DeploymentException` is reated in the validation phase, is it better that my current implementation that throws a `BuildException` ?

**Comment by machi1990** [comment]
> In PR #8576 I saw that a `DeploymentException` is reated in the validation phase, is it better that my current implementation that throws a `BuildException` ?

I would stick with `BuildException` or create a rather specific exception for that `DuplicateIDException` or something along those lines. The `DeploymentException` used there comes `javax.enterprise.inject.spi.DeploymentException` and I am not sure if it can be used here.

**Comment by FroMage** [comment]
Please note that @Sanne said that it's possible in ORM to have multiple `@Id` fields.

**Comment by FroMage** [comment]
But https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#identifiers-composite says that in this case `@IdClass` or `@EmbeddedId` must be used, unless @Sanne was referring to the deprecated variant that the docs mention?

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:87]: This should be recursive: it is possible that we have superclasses between the entity class and `PanacheEntity`.

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:90]:
```suggestion
                        "your class should extend PanacheEntityBase instead, or use the id provided by PanacheEntity", Collections.emptyList());
```

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:144]: Same remark about inheritance.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:147]:
```suggestion
                        "your class should extend PanacheMongoEntityBase instead, or use the id provided by PanageMongoEntity", Collections.emptyList());
```

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:235]: Same remark about inheritance.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:238]:
```suggestion
                        "your class should extend ReactivePanacheMongoEntityBase instead, or use the id provided by ReactivePanacheMongoEntity", Collections.emptyList());
```

**Review by FroMage — CHANGES_REQUESTED** [review]
Actually this does not detect whether we define more than one ID, but whether we define an ID _and_ extend `PanacheEntity` which also have an ID. Given that it's highly unlikely anyone will want to define a composite ID using both his own ID and the one from `PanacheEntity.id`, I think this works and doesn't need to check for `@IdClass`.

**Comment by Sanne** [comment]
> Actually this does not detect whether we define more than one ID, but whether we define an ID _and_ extend `PanacheEntity` which also have an ID. Given that it's highly unlikely anyone will want to define a composite ID using both his own ID and the one from `PanacheEntity.id`, I think this works and doesn't need to check for `@IdClass`.

Ok that's great then. Sorry for confusing things, just wanted to make sure the multiple @Id case was being taken into account - and yes it wouldn't be enough, you will need still an IdClass as well, I just meant to be careful as it's not as straight forward and I'd rather let Hibernate ORM throw a meaningfull error in such cases.

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java]: Done, I climb up the hierarchy of classes.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java]: Done, I climb up the hierarchy of classes.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java]: Done, I climb up the hierarchy of classes.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (FroMage) [inline extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java:227]:
```suggestion
    public static boolean isSubclassOf(IndexView index, ClassInfo info, DotName parentName) throws BuildException {
```

Given that this is only for superclasses (not interfaces) the name should make it clear.

**Review by FroMage — CHANGES_REQUESTED** [review]

**Comment by loicmathieu** [comment]
@FroMage I want to check the performance of this before merging it, so no hurry ;)

**Comment by loicmathieu** [comment]
@FroMage @machi1990 I move the check to the valdiation phase, this allow to divide by two the cost of the validation for MongoDB as it can check in one pass for both reactive and imperative entity.

The cost is around 500µs for the IT so it's highly acceptable ;)

Ready to be merged!

**Comment by FroMage** [comment]
There's a conflict :(

**Comment by loicmathieu** [comment]
@FroMage rebased, no more conflit!

**Review by FroMage — APPROVED** [review]

**Comment by loicmathieu** [comment]
Rebased on master to relaunch CI as the Cache native test timeouts.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:142]: Seems like the exception is not thrown anymore..

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:227]: Same here the exception is not thrown anymore

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: nitpick: We could maybe have an import of `DOTNAME_OBJECT ` and avoid this long line

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (machi1990) [inline extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java:346]:
```suggestion
            // this can happens if the parent is not inside the Jandex index
```

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (machi1990) [inline extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java:330]:
```suggestion
                throw new RuntimeException(); // this should not occur
```

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:306]:
```suggestion
            CombinedIndexBuildItem index)  {
```

No exception is thrown

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:162]:
```suggestion
            CombinedIndexBuildItem index)  {
```
We do not throw any exception

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [inline extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:214]: static import of `DOTNAME_OBJECT ` would be better instead of this very long line.

or we could import `JandexUtil` at the very least since it seems to be used everywhere.

**Review by machi1990 — CHANGES_REQUESTED** [review]
Hi @loicmathieu thanks for the great work as usual. I was about to merge this but noticed a couple of things, can you have a look?

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: Or we could just import `JandexUtil` since it seems to be used everywhere

Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: There is two JandexUtil, one from core deployment and one from Panache.
I think we need to rename the panache one PanacheJandexUtil to be able to import it.
I'll need @FroMage blessing on this rename as he is the original author ;)

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (loicmathieu) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java:346]: Ah funny, I always pronounce Jandex Yandex and this time I wrote it ;)

Yandex is the equivalent of Google in Russia ... integrated once in a website for Russia Maps ...

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:306]: It can be thrown by `io.quarkus.panache.common.deployment.JandexUtil.isSubclassOf` that delcares it as a checked exception.

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:162]: It can be thrown by `io.quarkus.panache.common.deployment.JandexUtil.isSubclassOf` that delcares it as a checked exception.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java]: Removed

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java]: Remove

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (loicmathieu) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java]: Fixed

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [reply extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:162]: Ah I might have missed it, sorry.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java
Comment (loicmathieu) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/JandexUtil.java]: Fixed

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (machi1990) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: Okay. Can be done in another PR I suppose. Not a biggie.

**Review by machi1990 — APPROVED** [review]

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (FroMage) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: If I get @stuartwdouglas 's blessing I would move most of our JandexUtil's methods into the deployment one, because it has so many useful methods for everyone, dealing with producing ASM from jandex-using code, that I'm sure other extensions had to reimplement a lot of it.

**Comment by loicmathieu** [comment]
JDK 8 CI failed with the Maven test (I saw it failed a lot lately).
JDK 11 CI failed with the Code Flow test (a timeout).

**Comment by machi1990** [comment]
The failure seems unrelated. Merging. 

@loicmathieu thanks.

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java
Comment (stuartwdouglas) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: Sounds reasonable.

Comment (FroMage) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:373]: OK thanks. I'll open a mail discussion before doing that PR, to make sure other `JandeUtil` classes don't already exist.

**Pull request merged**
